### PR TITLE
[clang][cas] Don't include PGO-related files in the include-tree when they are not used

### PR DESCRIPTION
**Pull request by akyrtzi**
PGO options are ignored for PCH invocations and may not be treated as dependencies from the build system, so don't add them in the include-tree. Otherwise we may get in a situation where such a file gets updated and an incremental build fails because of mismatched files in the include-tree of the PCH and the translation unit that uses it.

rdar://125629656

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci Please test LLVM

File: clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp [hunk 1]
```diff
@@ -952,7 +952,6 @@ void DependencyScanningWorker::computeDependenciesFromCompilerInvocation(
 
   // Adjust the invocation.
   auto &Frontend = Invocation->getFrontendOpts();
-  Frontend.ProgramAction = frontend::RunPreprocessorOnly;
```
Comment (benlangmuir) [inline clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp:944]: What's the reason for this change?

File: clang/include/clang/Tooling/DependencyScanning/ScanAndUpdateArgs.h [hunk 2]
```diff
@@ -40,6 +40,9 @@ void configureInvocationForCaching(CompilerInvocation &CI, CASOptions CASOpts,
                                    std::string RootID, std::string WorkingDir,
                                    bool ProduceIncludeTree);
 
+/// Resets some options that introduce dependencies unnecessarily.
+void removeUnnecessaryDependencies(CompilerInvocation &CI, bool ForModuleBuild);
```
Comment (benlangmuir) [inline clang/include/clang/Tooling/DependencyScanning/ScanAndUpdateArgs.h]: This should probably live in ModuleDepCollector.h or something similar; it's not specific to caching and we could potentially upstream this part of the change if we found annother reason to use this code.

File: clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp (hunk 1 above)
Comment (akyrtzi) [reply clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp:944]: This code path is only used for scanning via `-cc1depscan`, and it overrides the `frontend::GeneratePCH` program action, which the `removeUnnecessaryDependencies` function depends on being able to check for.

AFAICT the other scanning code-paths do not do this, so I assume it's not really necessary.

File: clang/include/clang/Tooling/DependencyScanning/ScanAndUpdateArgs.h (hunk 2 above)
Comment (akyrtzi) [reply clang/include/clang/Tooling/DependencyScanning/ScanAndUpdateArgs.h]: Yeah, good point, I'll do a NFC modification to add it, otherwise we'll have trouble with merge conflicts.

Comment (akyrtzi) [reply clang/include/clang/Tooling/DependencyScanning/ScanAndUpdateArgs.h]: See updated commit https://github.com/apple/llvm-project/pull/8564/commits/3b886bc00592494407de5de202a3c7dfadfcb960 (the other commit is an upstream commit that I cherry-picked to the PR, it's not in `next` yet).

**Review by jansvoboda11 — APPROVED** [review]

**Comment by akyrtzi** [comment]
@swift-ci Please test LLVM
